utils/estimated_histogram.hh

```diff
@@ -47,13 +47,376 @@
+        res.sample_sum = 0;
+
+        for (size_t i = 0; i < NUM_BUCKETS - 1; i++) {
+            auto& v = res.buckets[i];
+            v.upper_bound = get_bucket_limit(i + 1);
+            cummulative_count += buckets[i];
+            v.count = cummulative_count;
+            res.sample_sum += buckets[i] * v.upper_bound;
+        }
+        // The count serves as the infinite bucket
+        res.sample_count = cummulative_count + buckets[NUM_BUCKETS - 1];
+        res.sample_sum += buckets[NUM_BUCKETS - 1] * get_bucket_limit(NUM_BUCKETS - 1);
+        return res;
+    }
+
+    /*!
+     * \brief return the histogram buckets
+     */
+    const std::vector<uint64_t>& get_buckets() const {
+        return buckets;
+    }
+
+    void clear() {
+        std::fill(buckets.begin(), buckets.end(), 0);
+    }
+
+    /*!
+     * \brief Add an item to the histogram
+     * Increments the count of the bucket closest to n, rounding DOWN.
+     */
+    histogram_type& add(uint64_t n) {
+        buckets.at(find_bucket_pos(n))++;
+        return *this;
+    }
+
+    /*!
+     * \brief returns the smallest value that could have been added to this histogram
+     */
+    uint64_t min() const {
+        for (size_t i = 0; i < NUM_BUCKETS; i ++) {
+            if (buckets[i] > 0) {
+                return get_bucket_limit(i);
+            }
+        }
+        return 0;
+    }
+
+    /*!
+     * \brief returns the largest value that could have been added to this histogram.  If the histogram
+     * overflowed, returns UINT64_MAX.
+     */
+    uint64_t max() const {
+        if (buckets[NUM_BUCKETS - 1] > 0) {
+            return UINT64_MAX;
+        }
+        for (int i = NUM_BUCKETS - 1; i >= 0; i--) {
+            if (buckets[i] > 0) {
+                return get_bucket_limit(i);
+            }
+        }
+        return 0;
+    }
+
+    /*!
+     * \brief merge a histogram to the current one.
+     */
+    histogram_type& merge(const histogram_type& b) {
+        for (size_t i = 0; i < NUM_BUCKETS; i++) {
+            buckets[i] += b.get_buckets()[i];
+        }
+        return *this;
+    }
+
+    template<uint64_t A, uint64_t B, std::vector<uint64_t>::size_type C>
+    friend pseudo_floatingpoint_histogram<A, B, C> merge(pseudo_floatingpoint_histogram<A, B, C> a, const pseudo_floatingpoint_histogram<A, B, C>& b);
+
+    /*
+     * \brief returns the count in the given bucket
+     */
+    uint64_t get(size_t bucket) const {
+        return buckets[bucket];
+    }
+
+    /*!
+     * \brief get a histogram percentile
+     *
+     * returns the estimated value at given percentile
+     */
+    int64_t percentile(double perc) const {
+        assert(perc >= 0 && perc <= 1.0);
+        auto last_bucket = NUM_BUCKETS - 1;
+
+        auto c = count();
+
+        if (!c) {
+            return 0; // no data
+        }
+
+        auto pcount = int64_t(std::floor(c * perc));
+        int64_t elements = 0;
+        for (size_t i = 0; i < last_bucket; i++) {
```

Comment:
Aren't you missing the last bucket here? last_bucket is already NUM_BUCKETS - 1.

Reply:
I'll check again

Reply:
I've checked again looking at current implementation in estimated_histogram, it's on purpose so a no found value is found the result will be Max.